sympy/core/add.py

```diff
@@ -862,10 +862,33 @@ def _eval_as_leading_term(self, x):
 
         infinite = [t for t in expr.args if t.is_infinite]
 
-        expr = expr.func(*[t.as_leading_term(x) for t in expr.args]).removeO()
-        if not expr:
-            # simple leading term analysis gave us 0 but we have to send
-            # back a term, so compute the leading term (via series)
+        compute = False
+        leading_terms = [t.as_leading_term(x) for t in expr.args]
+        new_expr = expr.func(*leading_terms).removeO()
+
+        if not new_expr:
+            compute = True
+        elif new_expr is not S.NaN:
+            if new_expr.is_Add:
+                final_leading_terms = [t for t in new_expr.args]
+            else:
+                final_leading_terms = [new_expr]
+
+            if len(expr.args) != len(final_leading_terms):
+                coeff_added_terms = [t for t in leading_terms
+                                    if t not in final_leading_terms]
+                leading_exprs = [t.as_coeff_Mul()[1] for t in final_leading_terms]
+                canceled_terms = [t for t in coeff_added_terms
+                                 if t.as_coeff_Mul()[1] not in leading_exprs]
+                expr_sum = expr.func(*canceled_terms)
+                if expr_sum == S(0) and canceled_terms:
+                    compute = True
+
+        expr = new_expr
+
+        if compute:
+            # simple leading term analysis gave us cancelled terms whose sum is 0
+            # so compute the leading term (via series)
```

Comment:
I am confused on this step. if ``compute`` is ``True``  ``old.compute_leading_term(x)`` is returned, where is ``expr`` supposed to be used then.

Reply:
I will explain it to you.

`old.compute_leading_term(x)` is executed when the leading terms of expr.args cancel each other out perfectly. So `expr = expr.func(*[t.as_leading_term(x) for t in expr.args]).removeO()` turns expr into 0. 
But the current implementation does not check whether the leading terms cancelled each other partially.So instead of `expr`, `new_expr` is introduced `new_expr = expr.func(*[t.as_leading_term(x) for t in expr.args]).removeO()`.

Then number of arguments of `expr` and `new_expr` are checked. If they are not equal, i.e., `compute` is `True`, (old.compute_leading_term(x) is used. The following code will use `new_expr` now, not `expr`.

Reply:
> Then number of arguments of expr and new_expr are checked.

This is probably not helpful in general. The leading terms in `new_expr` have different orders. Only those of the lowest (or dominating) order matter. For example, assume that there are two arguments `a0, a1` of order 0 and two arguments `b0, b1`  of order 1. If the constant leading terms of `a0` and `a1` do not cancel, then the sum has order 0 and it does not matter at all if the leading terms of `b0` and `b1` cancel or not. On the other hand, if the constant terms cancel each other, then the first order terms become significant. But those of `b0` and `b1` do not suffice alone, the first order terms of `a0` and `a1` need to be computed as well.

So it seems to me that the following would be the proper procedure. Compute the list of all leading terms of the arguments. Filter the list omitting all but the dominating terms. Compute the sum of those terms and return it if it is not zero. Otherwise pass to `old.compute_leading_term()`.

Reply:
Hi @ArighnaIITG Any updates on this?

Reply:
@jksuom I am not being able to understand your point. Could you please explain?

Reply:
Maybe the filtering part is not clear. What I mean is this:
Make a list of the leading terms of the arguments. Then make another list of their orders as integers. Find the minimum of the orders. From the list of leading terms, remove all terms whose order is greater than the minimum. Compute the sum of the remaining leading terms all of which have the same order. If the sum is non-zero, then it is the leading term and can be returned. Otherwise pass to `old.compute_leading_term()`.